File: src/execution/runtimeRollbackEngine.ts
/**
 * Converted from Python: core/execution/runtime_rollback_engine.py
 * @generated — WebWeaveX python→javascript library port
 */

import * as py from "../runtime/pyCompat.js";

export function restoreRuntimeCheckpoint(checkpoint: any): any {
  return {"browser": py.pyDict(py.get(checkpoint, "browser", {})), "interaction": py.pyDict(py.get(checkpoint, "interaction", {})), "native": py.pyDict(py.get(checkpoint, "native", {})), "workflow": py.pyDict(py.get(checkpoint, "workflow", {})), "synchronization": py.pyDict(py.get(checkpoint, "synchronization", {})), "memory": py.pyDict(py.get(checkpoint, "memory", {})), "restored": true, "bounded": true};
}
export function rollbackRuntimeState(prior: any, current: any = null): any {
  var restored: any = restoreRuntimeCheckpoint(prior);
  return {"prior": prior, "current": py.or2(current, () => ({})), "restored_state": restored, "rolled_back": true, "replay_safe": true, "bounded": true};
}

File: lib/src/execution/runtime_rollback_engine.dart
/// Port of core/execution/runtime_rollback_engine.
Map<String, dynamic> _subMap(Map<String, dynamic> src, String key) {
  final dynamic v = src[key];
  if (v is Map) return <String, dynamic>{...v.cast<String, dynamic>()};
  return <String, dynamic>{};
}

Map<String, dynamic> restoreRuntimeCheckpoint(Map<String, dynamic> checkpoint) {
  return <String, dynamic>{
    'browser': _subMap(checkpoint, 'browser'),
    'interaction': _subMap(checkpoint, 'interaction'),
    'native': _subMap(checkpoint, 'native'),
    'workflow': _subMap(checkpoint, 'workflow'),
    'synchronization': _subMap(checkpoint, 'synchronization'),
    'memory': _subMap(checkpoint, 'memory'),
    'restored': true,
    'bounded': true,
  };
}

Map<String, dynamic> rollbackRuntimeState(
  Map<String, dynamic> prior, [
  Map<String, dynamic>? current,
]) {
  final Map<String, dynamic> restored = restoreRuntimeCheckpoint(prior);
  return <String, dynamic>{
    'prior': prior,
    'current': current ?? <String, dynamic>{},
    'restored_state': restored,
    'rolled_back': true,
    'replay_safe': true,
    'bounded': true,
  };
}
